src/ddspdrum/module.py

```diff
@@ -235,3 +236,166 @@ def play(self):
         amp = self.control_to_sample_rate(self.get_envelope())
         signal = self.fix_length(self.get_audio(), len(amp))
         return amp * signal
+
+
+class SVF(SynthModule):
+    """
+    A State Variable Filter that can do lowpass, highpass, bandpass, and bandreject
+    filtering. Allows modulation of the cutoff frequency and an adjustable resonance
+    parameter. Can self-oscillate to make a sine / cosine wave.
+    """
+
+    def __init__(
+        self,
+        mode: str = "LPF",
+        cutoff: float = 1000,
+        resonance: float = 0.707,
+        self_oscillate: bool = False,
+        sample_rate: int = SAMPLE_RATE,
+    ):
+        super().__init__()
+        self.sample_rate = sample_rate
+        self.mode = mode
+        self.cutoff = cutoff
+        self.resonance = resonance
+        self.self_oscillate = self_oscillate
+
+    def __call__(
+        self,
+        audio: np.ndarray,
+        cutoff_mod: np.ndarray = None,
+        cutoff_mod_amount: float = 0.0,
+    ) -> np.ndarray:
+        """
+        Process audio samples
+        """
+
+        h = np.zeros(2)
+        y = np.zeros_like(audio)
+
+        # Calculate initial coefficients
+        coeff0, coeff1, rho = self.calculate_coefficients(self.cutoff)
+
+        # Check if there is a filter cutoff envelope to apply
+        apply_modulation = False
+        if cutoff_mod is not None and cutoff_mod_amount != 0.0:
```

Comment:
I'm also slightly leery about mixing booleans and continuous values. Could we just use cutoff_mod_amount?

(Same reason being that if you are backprop'ing these parameters, the boolean will become a continuous in the range [0, 1])